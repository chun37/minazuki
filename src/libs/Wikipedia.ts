import WikiJS, { Page } from 'wikijs';

export class Wikipedia {
  protected client = WikiJS({
    apiUrl: 'https://ja.wikipedia.org/w/api.php',
  });

  /**
   * Fuzzy search an Wikipedia article
   * @param keyword Query string
   * @return Result Page object, I wonder why WikiJS didn't export it
   */
  public async search(keyword: string) {
    try {
      const { results } = await this.client.search(keyword);

      if (!results.length) {
        return null;
      }

      return (await this.client.page(results[0])) as Page & { pageid: string };
    } catch (e) {
      return null;
    }
  }
}
