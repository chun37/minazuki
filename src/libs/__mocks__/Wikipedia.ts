export const Wikipedia = jest.fn(() => ({
  search: jest.fn().mockResolvedValue({
    raw: {
      pageid: '123',
    },
    summary: jest.fn(() => 'summary'),
  }),
}));
