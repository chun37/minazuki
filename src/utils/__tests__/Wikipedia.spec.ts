import Wikipedia from '../Wikipedia';

describe('Wikipedia', () => {
  it ('searches an article of wikipedia with specific keyword', async () => {
    const wikipedia = new Wikipedia();
    const query     = 'Discord (ソフトウェア)';

    // The serach result is mutable...
    const result     = await wikipedia.search(query);
    const resultJson = JSON.stringify(result);

    expect(resultJson).toMatchSnapshot();
  });
});
