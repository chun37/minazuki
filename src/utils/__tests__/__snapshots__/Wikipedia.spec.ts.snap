// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Wikipedia searches an article of wikipedia with specific keyword 1`] = `"{\\"raw\\":{\\"pageid\\":3553115,\\"ns\\":0,\\"title\\":\\"Discord (ソフトウェア)\\",\\"contentmodel\\":\\"wikitext\\",\\"pagelanguage\\":\\"ja\\",\\"pagelanguagehtmlcode\\":\\"ja\\",\\"pagelanguagedir\\":\\"ltr\\",\\"touched\\":\\"2018-07-13T19:45:49Z\\",\\"lastrevid\\":69196379,\\"length\\":3290,\\"fullurl\\":\\"https://ja.wikipedia.org/wiki/Discord_(%E3%82%BD%E3%83%95%E3%83%88%E3%82%A6%E3%82%A7%E3%82%A2)\\",\\"editurl\\":\\"https://ja.wikipedia.org/w/index.php?title=Discord_(%E3%82%BD%E3%83%95%E3%83%88%E3%82%A6%E3%82%A7%E3%82%A2)&action=edit\\",\\"canonicalurl\\":\\"https://ja.wikipedia.org/wiki/Discord_(%E3%82%BD%E3%83%95%E3%83%88%E3%82%A6%E3%82%A7%E3%82%A2)\\"}}"`;
